Slather Cobertura report  integration  with Sonarqube  4.5.7

Has any one  integrated the cobertura.xml  file with the sonarqube  4.5.7

Using Slather 2.3.0
Using Xcode 7+

while running the  generated cobertura.xml against sonar the below error is generated 

 ERROR: Caused by: Error at line 3 of coverage report path to file /slather-report/cobertura.xml
 ERROR: Caused by: Unknown coverage version: Slather 2.3.0. This parser only handles version 1.

The first 3 lines in the file that is generated
<?xml version="1.0"?>
<!DOCTYPE coverage SYSTEM "http://cobertura.sourceforge.net/xml/coverage-04.dtd">
<coverage line-rate="0.0995231183910429" branch-rate="1.0000000000000000" lines-covered="960" lines-valid="9646" branches-covered="0" branches-valid="0" complexity="0.0" timestamp="1476702379" version="Slather 2.3.0">

Thanks in advance

I also get too,

> Look at SonarQube documentation to know the expected XML format.
> Caused by: java.lang.IllegalStateException: Unknown report version: Slather 2.4.7. This parser only handles version 1.

--- a/lib/slather.rb
+++ b/lib/slather.rb
@@ -12,6 +12,7 @@ require 'slather/coverage_service/simple_output'
 require 'slather/coverage_service/html_output'
 require 'slather/coverage_service/json_output'
 require 'slather/coverage_service/llvm_cov_output'
+require 'slather/coverage_service/sonarqube_xml_output'
 require 'cfpropertylist'
 
 module Slather

--- a/lib/slather/command/coverage_command.rb
+++ b/lib/slather/command/coverage_command.rb
@@ -13,6 +13,7 @@ class CoverageCommand < Clamp::Command
   option ["--simple-output", "-s"], :flag, "Output coverage results to the terminal"
   option ["--gutter-json", "-g"], :flag, "Output coverage results as Gutter JSON format"
   option ["--cobertura-xml", "-x"], :flag, "Output coverage results as Cobertura XML format"
+  option ["--sonarqube-xml", "-sq"], :flag, "Output coverage results as Cobertura XML format"
   option ["--llvm-cov", "-r"], :flag, "Output coverage as llvm-cov format"
   option ["--json"], :flag, "Output coverage results as simple JSON"
   option ["--html"], :flag, "Output coverage results as static html pages"
@@ -124,6 +125,8 @@ class CoverageCommand < Clamp::Command
       project.show_html = show?
     elsif json?
       project.coverage_service = :json
+    elsif sonarqube_xml?
+      project.coverage_service = :sonarqube_xml
     end
   end
 

--- a/lib/slather/coverage_service/sonarqube_xml_output.rb
+++ b/lib/slather/coverage_service/sonarqube_xml_output.rb
@@ -1,0 +1,61 @@
+require 'nokogiri'
+require 'date'
+
+module Slather
+  module CoverageService
+    module SonarqubeXmlOutput
+
+      def coverage_file_class
+        if input_format == "profdata"
+          Slather::ProfdataCoverageFile
+        else
+          Slather::CoverageFile
+        end
+      end
+      private :coverage_file_class
+
+      def post
+        cobertura_xml_report = create_xml_report(coverage_files)
+        store_report(cobertura_xml_report)
+      end
+
+      def store_report(report)
+        output_file = 'sonarqube-generic-coverage.xml'
+        if output_directory
+          FileUtils.mkdir_p(output_directory)
+          output_file = File.join(output_directory, output_file)
+        end
+        File.write(output_file, report.to_s)
+      end
+
+      def create_xml_report(coverage_files)
+        create_empty_xml_report
+        coverage_node = @doc.root
+        coverage_node['version'] = "1"
+
+        coverage_files.each do |coverage_file|
+          file_node = Nokogiri::XML::Node.new "file", @doc
+          file_node.parent = coverage_node
+          file_node['path'] = coverage_file.source_file_pathname_relative_to_repo_root.to_s
+          coverage_file.all_lines.each do |line|
+            if coverage_file.coverage_for_line(line)
+              line_node = Nokogiri::XML::Node.new "lineToCover", @doc
+              line_node['lineNumber'] = coverage_file.line_number_in_line(line)
+              line_node['covered'] = coverage_file.coverage_for_line(line) == 0 ? "false" : "true"
+              line_node.parent = file_node
+            end
+          end
+        end
+        @doc.to_xml
+      end
+
+      def create_empty_xml_report
+        builder = Nokogiri::XML::Builder.new do |xml|
+          xml.coverage
+        end
+        @doc = builder.doc
+      end
+
+    end
+  end
+end

--- a/lib/slather/project.rb
+++ b/lib/slather/project.rb
@@ -424,6 +424,8 @@ module Slather
         extend(Slather::CoverageService::HtmlOutput)
       when :json
         extend(Slather::CoverageService::JsonOutput)
+      when :sonarqube_xml
+        extend(Slather::CoverageService::SonarqubeXmlOutput)
       else
         raise ArgumentError, "`#{coverage_service}` is not a valid coverage service. Try `terminal`, `coveralls`, `gutter_json`, `cobertura_xml` or `html`"
       end

--- a/spec/fixtures/sonarqube-generic-coverage.xml
+++ b/spec/fixtures/sonarqube-generic-coverage.xml
@@ -1,0 +1,93 @@
+<?xml version="1.0"?>
+<coverage version="1">
+  <file path="spec/fixtures/fixtures/fixtures.m">
+    <lineToCover lineNumber="14" covered="true"/>
+    <lineToCover lineNumber="15" covered="true"/>
+    <lineToCover lineNumber="16" covered="true"/>
+    <lineToCover lineNumber="19" covered="false"/>
+    <lineToCover lineNumber="20" covered="false"/>
+    <lineToCover lineNumber="21" covered="false"/>
+  </file>
+  <file path="spec/fixtures/fixtures/more_files/Branches.m">
+    <lineToCover lineNumber="14" covered="true"/>
+    <lineToCover lineNumber="15" covered="true"/>
+    <lineToCover lineNumber="16" covered="true"/>
+    <lineToCover lineNumber="17" covered="true"/>
+    <lineToCover lineNumber="18" covered="true"/>
+    <lineToCover lineNumber="19" covered="false"/>
+    <lineToCover lineNumber="20" covered="false"/>
+    <lineToCover lineNumber="21" covered="true"/>
+    <lineToCover lineNumber="22" covered="true"/>
+    <lineToCover lineNumber="23" covered="true"/>
+    <lineToCover lineNumber="24" covered="true"/>
+    <lineToCover lineNumber="25" covered="true"/>
+    <lineToCover lineNumber="26" covered="true"/>
+    <lineToCover lineNumber="27" covered="true"/>
+    <lineToCover lineNumber="28" covered="true"/>
+    <lineToCover lineNumber="29" covered="false"/>
+    <lineToCover lineNumber="30" covered="false"/>
+    <lineToCover lineNumber="31" covered="false"/>
+    <lineToCover lineNumber="32" covered="false"/>
+    <lineToCover lineNumber="33" covered="false"/>
+    <lineToCover lineNumber="34" covered="false"/>
+    <lineToCover lineNumber="35" covered="false"/>
+    <lineToCover lineNumber="36" covered="false"/>
+    <lineToCover lineNumber="37" covered="false"/>
+    <lineToCover lineNumber="38" covered="false"/>
+    <lineToCover lineNumber="39" covered="false"/>
+    <lineToCover lineNumber="40" covered="false"/>
+    <lineToCover lineNumber="41" covered="false"/>
+    <lineToCover lineNumber="42" covered="false"/>
+    <lineToCover lineNumber="43" covered="false"/>
+  </file>
+  <file path="spec/fixtures/fixturesTests/BranchesTests.m">
+    <lineToCover lineNumber="18" covered="true"/>
+    <lineToCover lineNumber="19" covered="true"/>
+    <lineToCover lineNumber="20" covered="true"/>
+    <lineToCover lineNumber="21" covered="true"/>
+    <lineToCover lineNumber="23" covered="true"/>
+    <lineToCover lineNumber="24" covered="true"/>
+    <lineToCover lineNumber="25" covered="true"/>
+    <lineToCover lineNumber="26" covered="true"/>
+    <lineToCover lineNumber="28" covered="true"/>
+    <lineToCover lineNumber="29" covered="true"/>
+    <lineToCover lineNumber="30" covered="true"/>
+    <lineToCover lineNumber="31" covered="true"/>
+    <lineToCover lineNumber="33" covered="true"/>
+    <lineToCover lineNumber="34" covered="true"/>
+    <lineToCover lineNumber="35" covered="true"/>
+    <lineToCover lineNumber="36" covered="true"/>
+  </file>
+  <file path="spec/fixtures/fixturesTests/fixturesTests.m">
+    <lineToCover lineNumber="20" covered="true"/>
+    <lineToCover lineNumber="21" covered="true"/>
+    <lineToCover lineNumber="22" covered="true"/>
+    <lineToCover lineNumber="23" covered="true"/>
+    <lineToCover lineNumber="26" covered="true"/>
+    <lineToCover lineNumber="27" covered="true"/>
+    <lineToCover lineNumber="28" covered="true"/>
+    <lineToCover lineNumber="29" covered="true"/>
+    <lineToCover lineNumber="32" covered="true"/>
+    <lineToCover lineNumber="33" covered="true"/>
+    <lineToCover lineNumber="34" covered="true"/>
+    <lineToCover lineNumber="35" covered="true"/>
+    <lineToCover lineNumber="38" covered="true"/>
+    <lineToCover lineNumber="39" covered="true"/>
+    <lineToCover lineNumber="40" covered="true"/>
+    <lineToCover lineNumber="41" covered="true"/>
+    <lineToCover lineNumber="42" covered="true"/>
+  </file>
+  <file path="spec/fixtures/fixturesTests/peekaviewTests&#x1F4A3;.m">
+    <lineToCover lineNumber="18" covered="true"/>
+    <lineToCover lineNumber="19" covered="true"/>
+    <lineToCover lineNumber="20" covered="true"/>
+    <lineToCover lineNumber="21" covered="true"/>
+    <lineToCover lineNumber="24" covered="true"/>
+    <lineToCover lineNumber="25" covered="true"/>
+    <lineToCover lineNumber="26" covered="true"/>
+    <lineToCover lineNumber="27" covered="true"/>
+    <lineToCover lineNumber="30" covered="true"/>
+    <lineToCover lineNumber="31" covered="true"/>
+    <lineToCover lineNumber="32" covered="true"/>
+  </file>
+</coverage>

--- a/spec/slather/coverage_service/sonarqube_xml_spec.rb
+++ b/spec/slather/coverage_service/sonarqube_xml_spec.rb
@@ -1,0 +1,46 @@
+require File.join(File.dirname(__FILE__), '..', '..', 'spec_helper')
+require 'json'
+
+describe Slather::CoverageService::SonarqubeXmlOutput do
+
+  let(:fixtures_project) do
+    proj = Slather::Project.open(FIXTURES_PROJECT_PATH)
+    proj.build_directory = TEMP_DERIVED_DATA_PATH
+    proj.input_format = "profdata"
+    proj.coverage_service = "sonarqube_xml"
+    proj.configure
+    proj
+  end
+
+  describe '#coverage_file_class' do
+    it "should return CoverageFile" do
+      expect(fixtures_project.send(:coverage_file_class)).to eq(Slather::ProfdataCoverageFile)
+    end
+  end
+
+  describe '#post' do
+    it "should create an XML report spanning all coverage files" do
+      fixtures_project.post
+
+      file = File.open(FIXTURES_SONARQUBE_XML_PATH)
+      fixture_xml_doc = Nokogiri::XML(file)
+      file.close
+
+      file = File.open('sonarqube-generic-coverage.xml')
+      current_xml_doc = Nokogiri::XML(file)
+      file.close
+
+      expect(EquivalentXml.equivalent?(current_xml_doc, fixture_xml_doc)).to be_truthy
+    end
+
+    it "should create an XML report in the given output directory" do
+      fixtures_project.output_directory = "./output"
+      fixtures_project.post
+
+      filepath = "#{fixtures_project.output_directory}/sonarqube-generic-coverage.xml"
+      expect(File.exists?(filepath)).to be_truthy
+
+      FileUtils.rm_rf(fixtures_project.output_directory)
+    end
+  end
+end

--- a/spec/spec_helper.rb
+++ b/spec/spec_helper.rb
@@ -13,6 +13,7 @@ require 'equivalent-xml'
 
 
 FIXTURES_XML_PATH = File.join(File.dirname(__FILE__), 'fixtures/cobertura.xml')
+FIXTURES_SONARQUBE_XML_PATH = File.join(File.dirname(__FILE__), 'fixtures/sonarqube-generic-coverage.xml')
 FIXTURES_JSON_PATH = File.join(File.dirname(__FILE__), 'fixtures/report.json')
 FIXTURES_LLCOV_PATH = File.join(File.dirname(__FILE__), 'fixtures/report.llcov')
 FIXTURES_GUTTER_JSON_PATH = File.join(File.dirname(__FILE__), 'fixtures/gutter.json')

